File: edukate-auth/src/main/java/io/github/sanyavertolet/edukate/auth/services/AuthCookieService.java
package io.github.sanyavertolet.edukate.auth.services;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.web.server.Cookie;
import org.springframework.http.HttpCookie;
import org.springframework.http.ResponseCookie;
import org.springframework.http.ResponseEntity;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Mono;

@Service
public class AuthCookieService {
    private static final String AUTHORIZATION_COOKIE = "X-Auth";

    @Value("${jwt.expirationSeconds}")
    private long expirationTimeSeconds;

    private ResponseCookie createTokenCookie(String token) {
        return ResponseCookie.from(AUTHORIZATION_COOKIE, token)
                .httpOnly(true)
                .secure(true)
                .path("/")
                .maxAge(expirationTimeSeconds)
                .sameSite(Cookie.SameSite.STRICT.toString())
                .build();
    }

    private ResponseCookie createExpiredTokenCookie() {
        return ResponseCookie.from(AUTHORIZATION_COOKIE, "expired")
                .httpOnly(true)
                .secure(true)
                .path("/")
                .maxAge(0)
                .sameSite(Cookie.SameSite.STRICT.toString())
                .build();
    }

    public Mono<String> ejectToken(ServerWebExchange exchange) {
        return Mono.fromCallable(exchange::getRequest)
                .map(ServerHttpRequest::getCookies)
                .mapNotNull(cookies -> cookies.getFirst(AUTHORIZATION_COOKIE))
                .map(HttpCookie::getValue);
    }

    public Mono<ResponseEntity<Void>> respondWithToken(String token) {
        return Mono.justOrEmpty(token)
                .map(this::createTokenCookie)
                .map(cookie ->
                        ResponseEntity.noContent().header("Set-Cookie", cookie.toString()).build()
                );
    }

    public Mono<ResponseEntity<Void>> respondWithExpiredToken() {
        return Mono.fromCallable(this::createExpiredTokenCookie).map(cookie ->
                ResponseEntity.noContent().header("Set-Cookie", cookie.toString()).build()
        );
    }
}

File: edukate-auth/src/main/kotlin/io/github/sanyavertolet/edukate/auth/services/AuthCookieService.kt
package io.github.sanyavertolet.edukate.auth.services

import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.web.server.Cookie
import org.springframework.core.env.Environment
import org.springframework.http.ResponseCookie
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toMono

@Service
class AuthCookieService(
    @param:Value($$"${auth.jwt.expirationSeconds}") private val expirationTimeSeconds: Long,
    @param:Value($$"${auth.hostname}") private val hostname: String,
    environment: Environment,
) {
    private val isSecure = !environment.matchesProfiles("dev")

    private fun createTokenCookie(token: String): ResponseCookie =
        ResponseCookie.from(AUTHORIZATION_COOKIE, token)
            .httpOnly(true)
            .secure(isSecure)
            .domain(hostname)
            .path("/")
            .maxAge(expirationTimeSeconds)
            .sameSite(Cookie.SameSite.STRICT.toString())
            .build()

    private fun createExpiredTokenCookie(): ResponseCookie =
        ResponseCookie.from(AUTHORIZATION_COOKIE, EXPIRED)
            .httpOnly(true)
            .secure(isSecure)
            .domain(hostname)
            .path("/")
            .maxAge(0)
            .sameSite(Cookie.SameSite.STRICT.toString())
            .build()

    fun ejectToken(exchange: ServerWebExchange): Mono<String> =
        exchange.request.cookies.getFirst(AUTHORIZATION_COOKIE)?.value.toMono()

    fun respondWithToken(token: String): Mono<ResponseEntity<Void>> =
        Mono.fromCallable { createTokenCookie(token) }
            .map { ResponseEntity.noContent().header(SET_COOKIE_HEADER_NAME, it.toString()).build() }

    fun respondWithExpiredToken(): Mono<ResponseEntity<Void>> =
        Mono.fromCallable { createExpiredTokenCookie() }
            .map { ResponseEntity.noContent().header(SET_COOKIE_HEADER_NAME, it.toString()).build() }

    companion object {
        private const val EXPIRED = "expired"
        private const val SET_COOKIE_HEADER_NAME = "Set-Cookie"
        private const val AUTHORIZATION_COOKIE = "X-Auth"
    }
}
